pytorch_lightning/trainer/trainer.py

```diff
@@ -615,6 +616,7 @@ def run_train(self) -> None:
 
                 if self.should_stop:
                     if met_min_epochs and met_min_steps:
+                        self.train_loop.on_train_end()
```

Comment:
shall we also remove this from `train_loop.on_train_end` ? https://github.com/PyTorchLightning/pytorch-lightning/blob/b7a22ba046ba57072a71b12d16caff000e66f798/pytorch_lightning/trainer/training_loop.py#L114-L118

as this is one of the possible sources of bugs with the attempt at graceful shutdown

Reply:
Yes. I have a PR in progress for this. Need to finish it. This is part of what causes https://github.com/PyTorchLightning/pytorch-lightning/issues/6470

Reply:
@ananthsub , for this case, train does not have error (just early stopping), should we run `train_loop.on_train_end` to perform checkpointing?

@camruta , for why here I use `train_loop.on_train_end`, but for the case exception, I use `accelerator.on_train_end`, is because trying to avoid checkpointing when there is error for some ranks but not all (hence hang in broadcast).


To avoid divergence (normal end vs. error end), do you think we actually should remove the problematic checkpointing logic in `training_loop.on_train_end`. But there is also risk of adding logics in `training_loop.on_train_end` which could potentially need all ranks to be involved. How do we enforce that.

Reply:
open to discussion